invisible_cities/cities/components.py

```diff
@@ -519,3 +523,54 @@ def waveform_integrator(limits):
     def integrate_wfs(wfs):
         return cf.spaced_integrals(wfs, limits)[:, ::2]
     return integrate_wfs
+
+
+# Compound components
+def compute_and_write_pmaps(detector_db, run_number,
+                  s1_lmax, s1_lmin, s1_rebin_stride, s1_stride, s1_tmax, s1_tmin,
+                  s2_lmax, s2_lmin, s2_rebin_stride, s2_stride, s2_tmax, s2_tmin, thr_sipm_s2,
+                  h5out, compression, sipm_rwf_to_cal=None):
+
+    # Filter events without signal over threshold
+    indices_pass    = fl.map(check_nonempty_indices,
+                             args = ("s1_indices", "s2_indices"),
+                             out = "indices_pass")
+    empty_indices   = fl.count_filter(bool, args = "indices_pass")
+
+    # Build the PMap
+    compute_pmap     = fl.map(build_pmap(detector_db, run_number,
+                                         s1_lmax, s1_lmin, s1_rebin_stride, s1_stride, s1_tmax, s1_tmin,
+                                         s2_lmax, s2_lmin, s2_rebin_stride, s2_stride, s2_tmax, s2_tmin, thr_sipm_s2),
+                              args = ("ccwfs", "s1_indices", "s2_indices", "sipm"),
+                              out  = "pmap")
+
+    # Filter events with zero peaks
+    pmaps_pass      = fl.map(check_empty_pmap, args = "pmap", out = "pmaps_pass")
+    empty_pmaps     = fl.count_filter(bool, args = "pmaps_pass")
+
+    # Define writers...
+    write_pmap_         = pmap_writer         (h5out, compression=compression)
+    write_indx_filter_  = event_filter_writer (h5out, "s12_indices", compression=compression)
+    write_pmap_filter_  = event_filter_writer (h5out, "empty_pmap" , compression=compression)
+
+    # ... and make them sinks
+    write_pmap         = sink(write_pmap_        , args=(        "pmap", "event_number"))
+    write_indx_filter  = sink(write_indx_filter_ , args=("event_number", "indices_pass"))
+    write_pmap_filter  = sink(write_pmap_filter_ , args=("event_number",   "pmaps_pass"))
+
+    fn_list = (indices_pass,
+               fl.branch(write_indx_filter),
+               empty_indices.filter,
+               sipm_rwf_to_cal,
+               compute_pmap,
+               pmaps_pass,
+               fl.branch(write_pmap_filter),
+               empty_pmaps.filter,
+               fl.branch(write_pmap))
+
+    # remove sipm_rwf_to_cal if it is not set
```

Comment:
`compute_pmaps = pipe(*filter(None, fn_list))` ?

1. Filter without a predicate uses the boolean value of the elements themselves. I find the signal to noise ratio in the genexpr version rather low.
2. `this_name = fn(this_name)` is a style that you should (mildly) avoid, as the reader might see the first binding of `this _name` and not notice the rebinding later on. It's not *vitally* important, but it does help a bit, for reasons similar to those for which we prefer to avoid mutation if sanely possible.
3. There is absolutely no point in creating an intermediate list out of the lazy object (regardless of whether that lazy object is a generator expression (as you had) or a filter (as I am suggesting): any *iterable* (that includes everything that is lazy) can be used with the arg-expanding `*`.

The comment you wrote is vitally important. If you switch to using `filter` (and even if you don't), I suggest you change it to *filter out simp_rwf_to_cal if it is not set*.

Reply:
Grrrr! GitHub meta-question: is there a way of explicitly selecting the lines that will be shown above these comments? I never remember what GitHub is going to choose, and whatever it does't it clearly doesn't agree with my natural expectations, so, most of the time, the code it shows is not the code I'm talking about. I'm talking about the code on lines 571-571 plus a few.

Reply:
> `compute_pmaps = pipe(*filter(None, fn_list))` ?
> 
> 1. Filter without a predicate uses the boolean value of the elements themselves. I find the signal to noise ratio in the genexpr version rather low.

You are totally right, I have already updated the code with that.

> 2. `this_name = fn(this_name)` is a style that you should (mildly) avoid, as the reader might see the first binding of `this _name` and not notice the rebinding later on. It's not _vitally_ important, but it does help a bit, for reasons similar to those for which we prefer to avoid mutation if sanely possible.

Make sense too, not a problem with the new version.

> 3. There is absolutely no point in creating an intermediate list out of the lazy object (regardless of whether that lazy object is a generator expression (as you had) or a filter (as I am suggesting): any _iterable_ (that includes everything that is lazy) can be used with the arg-expanding `*`.

You are right, I was using `list` before because I had a line with `print(fn_list)` to check the `None` was being removed.

> The comment you wrote is vitally important. If you switch to using `filter` (and even if you don't), I suggest you change it to _filter out simp_rwf_to_cal if it is not set_.

Updated too.

Reply:
> Grrrr! GitHub meta-question: is there a way of explicitly selecting the lines that will be shown above these comments? I never remember what GitHub is going to choose, and whatever it does't it clearly doesn't agree with my natural expectations, so, most of the time, the code it shows is not the code I'm talking about. I'm talking about the code on lines 571-571 plus a few.

This I don't know, I'm not familiar either with that part of github's interface...

Reply:
> I was using `list` before because I had a line with `print(fn_list)` to check the `None` was being removed.

Yes, this single-use aspect of lazy things is a bit annoying, especially for interactivity and debugging.